Changed signature of Engine::updateReferences

sguazt opened the pull request:
Hello,

Since I'm working to an extension (i.e., derived class) of the `Engine` class, I need that `Engine::updateReferences` be `virtual`.

Furthermore, the `const` specifier causes me some trouble. I've noted that if I remove it, the compilation is still OK.

Can you consider to merge this request?

Thank you.

Best,

Marco

jcrada commented:
Hi Marco,

I am sorry for the late reply.

Why did you remove the const of the method? What do you intend to do in updateReferences()?

sguazt commented:
Hi Juan,

I'm developing an ANFIS model (I hope to be able to make a pull request soon...) and to do so I'm extending the `fl::Engine` class with a derived class, currently named `fl::anfis::Engine`.
The ANFIS model is made of `Node` objects (stored as members of `fl::anfis::Engine`) and each object contains a reference (a pointer) to the  `fl::anfis::Engine` class who owns it.

I  call `updateReferences` in `operator=` of `fl::anfis::Engine` to refresh references of copied objects (variables, rules, nodes, ...).

In the overridden `fl::anfis::Engine::updateReferences` method, in addition to call the base class method, I reset the reference to the current `fl::anfis::Engine` object for each ANFIS `Node` object.  For instance:

```c++
namespace fl { namespace anfis {

void Engine::updateReferences()
{
    BaseType::updateReferences();

    for (std::size_t i = 0,
                     n = inputNodes_.size();
         i < n;
         ++i)
    {
        this->inputNodes_[i]->setEngine(this);
    }
   ...
}
...
}} // Namespace fl::anfis
```
So, this is the reason of why I need to remove the `const`ness from `updateReferences`.
Regarding the `virtual` specifier, I need it since if `fl::anfis::Engine` is passed to a function as `fl::Engine*` (or `fl::Engine&`) and `updateReferences` is called, without the `virtual` specifier the version of `updateReferences` that will be used is the wrong one (i.e., `fl::Engine::updateReferences` instead of `fl::anfis::Engine::updateReferences`).

If do you see something wrong, let me know. I'm trying to understand *fuzzylite* by digging into the code and I could have missed something :)

Thanks,

Marco

jcrada commented:
Hi Marco,

I do not think is a good idea to override updateReferences because it is used in the copy constructor. This is something I need to address in other classes in fuzzylite too, as it is not safe to override such methods. Just think that allowing you to override it, when you use the copy constructor of your engine, it will attempt to set up your `anfis::Engine` before you have done it yourself because the `fl::Engine` copy constructor is called first. Perhaps it would be better to name your method differently.

sguazt commented:
OK, I've renamed my `updateReferences` into `updateAnfisReferences` and I've changed `operator=` accordingly.
It seems to work, without memleaks.

I close this request.

Thanks,

Marco
